src/main/java/com/io/image/manager/service/operations/ImageOperationParser.java

```diff
@@ -7,50 +7,90 @@
 
-        return new ConversionInfo("png",0);
+        String fileExtension = filename.split("\\.")[1];
+
+        if (query == null) {
+            return new ConversionInfo(fileExtension,1);
+        }
+        else if (query.charAt(0) == 'c' || query.charAt(0) == 'q') {
+            int rate = Integer.parseInt(query.split("&")[0].split("=")[1]);
+            return new ConversionInfo(fileExtension, rate);
+        }
+        return new ConversionInfo(fileExtension,1);
     }
 
-    public static List<ImageOperation> parse(String query) {
+    public static List<ImageOperation> parseAndGetOperationList(String query) {
 
         if (query == null) {
             return Collections.emptyList();
         }
 
-        List<ImageOperation> imageOperations = new ArrayList<>();
+        List<String> params = new ArrayList<>(Arrays.asList(query.split("&")));
+
+        boolean isWatermark = isWatermarking(params);
+
+        String[] args = params.get(0).split("=");
+        if ("p".equals(args[0]) || "q".equals(args[0])) {
+            params.remove(0);
+        }
+
+        return parse(params, isWatermark);
+    }
 
-        String[] params = query.split("&");
+    private static List<ImageOperation> parse(List<String> params, boolean isWatermark) {
 
+        List<ImageOperation> imageOperations = new ArrayList<>();
         ImageOperation imageOperation = null;
 
-        for (String param : params) {
-            String[] args = param.split("=");
-            String key = args[0];
-            String value = args[1];
+        if (params.size() > 0) {
+            for (String param : params) {
+                String[] args = param.split("=");
+                String key = args[0];
+                String value = args[1];
+
+                if ("op".equals(key)) {
+                    if (imageOperation != null) {
+                        imageOperations.add(imageOperation);
+                    }
 
-            if ("op".equals(key)) {
-                if (imageOperation != null) {
-                    imageOperations.add(imageOperation);
+                    imageOperation = createImageOperation(value);
+                    continue;
                 }
 
-                imageOperation = createImageOperation(value);
-                continue;
+                Objects.requireNonNull(imageOperation).addArgument(key, value);
             }
-
-            Objects.requireNonNull(imageOperation).addArgument(key, value);
+            imageOperations.add(imageOperation);
+        }
+        // This operation must be last because other operations can remove it
+        if (isWatermark) {
+            imageOperation = createImageOperation("watermark");
+            imageOperations.add(imageOperation);
         }
-        imageOperations.add(imageOperation);
-
         return imageOperations;
     }
 
+    private static boolean isWatermarking(List<String> params) {
```

Comment:
I don't like this function as the name does not suggest that any params will be removed from params. The checking is also wild, I know that regexes are pure evil but maybe it will look cleaner? Also applying regex to the whole query string would be a good idea for checking  that it is valid.